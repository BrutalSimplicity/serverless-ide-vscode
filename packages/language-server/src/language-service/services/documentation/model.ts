export const LIST_TYPE: 'List' = 'List';
export const MAP_TYPE: 'Map' = 'Map';
export const STRING_TYPE = 'String';
export const INTEGER_TYPE = 'Integer';
export const BOOLEAN_TYPE = 'Boolean';

export type PrimitiveType =
	| typeof STRING_TYPE
	| typeof INTEGER_TYPE
	| typeof BOOLEAN_TYPE;

export type ItemType = {
	Type?: typeof LIST_TYPE | typeof MAP_TYPE | string;
	ItemType?: string;
	PrimitiveType?: PrimitiveType;
	PrimitiveItemType?: typeof STRING_TYPE | typeof INTEGER_TYPE;
};

export type Property = {
	Documentation: string;
	Required: boolean;
	UpdateType?: 'Mutable' | 'Immutable' | 'Conditional';
	DuplicatesAllowed?: boolean;
} & ItemType;

export type Attribute = ItemType;

export interface PropertySpecification {
	Documentation: string;
	Properties: {
		[key: string]: Property;
	};
}

export interface ResourceSpecification {
	Documentation: string;
	Attributes: {
		[key: string]: Attribute;
	};
	Properties: {
		[key: string]: Property;
	};
}

export interface Specification {
	PropertyTypes: {
		[key: string]: PropertySpecification;
	};
	ResourceTypes: {
		[key: string]: ResourceSpecification;
	};
}
