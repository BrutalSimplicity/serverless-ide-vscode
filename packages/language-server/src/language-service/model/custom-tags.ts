import { ReferenceType } from "./references"

export type TagKind = "sequence" | "scalar" | "mapping"

export interface CustomTag {
	type?: ReferenceType
	tag: string
	kind: TagKind
	propertyName: string
}

export const CUSTOM_TAGS: CustomTag[] = [
	{
		tag: "!Base64",
		kind: "sequence",
		propertyName: "Fn::Base64"
	},
	{
		tag: "!Cidr",
		kind: "sequence",
		propertyName: "Fn::Cidr"
	},
	// logical
	{
		tag: "!And",
		kind: "sequence",
		propertyName: "Fn::And"
	},
	{
		tag: "!If",
		kind: "sequence",
		propertyName: "Fn::If"
	},
	{
		tag: "!Not",
		kind: "sequence",
		propertyName: "Fn::Not"
	},
	{
		tag: "!Or",
		kind: "sequence",
		propertyName: "Fn::Or"
	},
	{
		tag: "!Equals",
		kind: "sequence",
		propertyName: "Fn::Equals"
	},
	// getters
	{
		type: ReferenceType.FIND_IN_MAP,
		tag: "!FindInMap",
		kind: "sequence",
		propertyName: "Fn::FindInMap"
	},
	{
		type: ReferenceType.GET_ATT,
		tag: "!GetAtt",
		kind: "scalar",
		propertyName: "Fn::GetAtt"
	},
	{
		tag: "!GetAZs",
		kind: "scalar",
		propertyName: "Fn::GetAZs"
	},
	{
		tag: "!ImportValue",
		kind: "mapping",
		propertyName: "Fn::ImportValue"
	},
	{
		tag: "!Join",
		kind: "sequence",
		propertyName: "Fn::Join"
	},
	{
		tag: "!Select",
		kind: "sequence",
		propertyName: "Fn::Select"
	},
	{
		tag: "!Split",
		kind: "sequence",
		propertyName: "Fn::Split"
	},
	{
		type: ReferenceType.SUB,
		tag: "!Sub",
		kind: "sequence",
		propertyName: "Fn::Sub"
	},
	{
		tag: "!Transform",
		kind: "mapping",
		propertyName: "Fn::Transform"
	},
	// ref
	{
		type: ReferenceType.REF,
		tag: "!Ref",
		kind: "scalar",
		propertyName: "Ref"
	},
	{
		type: ReferenceType.DEPENDS_ON,
		tag: "",
		kind: "sequence",
		propertyName: "DependsOn"
	}
]
