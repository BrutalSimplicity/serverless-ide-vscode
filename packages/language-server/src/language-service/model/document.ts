export enum DocumentType {
	CLOUD_FORMATION = "CLOUD_FORMATION",
	SAM = "SAM",
	SERVERLESS_FRAMEWORK = "SERVERLESS_FRAMEWORK",
	UNKNOWN = "UNKNOWN"
}
