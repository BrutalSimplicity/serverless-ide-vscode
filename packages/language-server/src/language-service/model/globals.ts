import { Segment } from "vscode-json-languageservice"

export const API: "Api" = "Api"
export const FUNCTION: "Function" = "Function"
export const SIMPLE_TABLE: "SimpleTable" = "SimpleTable"

export type GlobalKeyType = typeof API | typeof FUNCTION | typeof SIMPLE_TABLE

export const GlobalKeys = {
	[API]: {
		resourceType: "AWS::Serverless::Api"
	},
	[FUNCTION]: {
		resourceType: "AWS::Serverless::Function"
	},
	[SIMPLE_TABLE]: {
		resourceType: "AWS::Serverless::SimpleTable"
	}
}

export interface GlobalConfigItem {
	resourceType: string
	properties: Segment[]
}

export interface GlobalsConfig {
	[API]: GlobalConfigItem
	[FUNCTION]: GlobalConfigItem
	[SIMPLE_TABLE]: GlobalConfigItem
}

const KEYS = [API, FUNCTION, SIMPLE_TABLE]

export const isEmpty = (globalsConfig: GlobalsConfig): boolean => {
	return KEYS.every(key => globalsConfig[key].properties.length === 0)
}
