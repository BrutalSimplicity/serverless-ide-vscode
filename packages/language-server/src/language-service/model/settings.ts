export enum ValidationProvider {
	default = "default",
	"cfn-lint" = "cfn-lint"
}

export interface CFNLintExtensionSettings {
	path?: string
	appendRules?: string[]
	ignoreRules?: string[]
	overrideSpecPath?: string
}

export interface ExtensionSettings {
	serverlessIDE: {
		validationProvider?: ValidationProvider
		cfnLint?: CFNLintExtensionSettings
		validate?: boolean
		hover?: boolean
		completion?: boolean
	}
	http: {
		proxy: string
		proxyStrictSSL: boolean
	}
}

export interface CFNLintSettings {
	path: string
	appendRules: string[]
	ignoreRules: string[]
	overrideSpecPath?: string
}

export interface LanguageSettings {
	validate: boolean
	hover: boolean
	completion: boolean
	customTags: string[]
	validationProvider: ValidationProvider
	cfnLint: CFNLintSettings
	workspaceRoot: string
}

export const getDefaultLanguageSettings = (): LanguageSettings => ({
	validate: true,
	hover: true,
	completion: true,
	customTags: [],
	validationProvider: ValidationProvider["cfn-lint"],
	cfnLint: {
		path: "cfn-lint",
		appendRules: [],
		ignoreRules: []
	},
	workspaceRoot: ""
})
