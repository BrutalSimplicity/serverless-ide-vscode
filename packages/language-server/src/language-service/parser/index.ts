"use strict"

import {
	ArrayASTNode,
	ASTNode,
	BooleanASTNode,
	ErrorCode,
	JSONDocument,
	NullASTNode,
	NumberASTNode,
	ObjectASTNode,
	PropertyASTNode,
	StringASTNode
} from "./jsonParser"

import * as nls from "vscode-nls"
const localize = nls.loadMessageBundle()

import { Schema, Type } from "js-yaml"
import * as Yaml from "yaml-ast-parser"

import { GlobalsConfig } from "../model/globals"
import {
	getLineStartPositions,
	getPosition
} from "../utils/documentPositionCalculator"
import { collectGlobals, getDefaultGlobalsConfig } from "./globals"
import { parseYamlBoolean } from "./scalar-type"

export interface Problem {
	message: string
	location: {
		start: number
		end: number
	}
	code: ErrorCode
}

export class SingleYAMLDocument extends JSONDocument {
	root: ASTNode
	errors: Problem[]
	warnings: Problem[]
	globalsConfig: GlobalsConfig
	private lines: number[]

	constructor(lines: number[]) {
		super(null, [])
		this.globalsConfig = getDefaultGlobalsConfig()
		this.lines = lines
		this.root = null
		this.errors = []
		this.warnings = []
	}

	getSchemas(schema, doc, node: ASTNode) {
		const matchingSchemas = []
		doc.validate(schema, matchingSchemas, node.start)
		return matchingSchemas
	}

	getNodeFromOffset(offset: number): ASTNode {
		return this.getNodeFromOffsetEndInclusive(offset)
	}

	private getNodeByIndent = (
		lines: number[],
		offset: number,
		node: ASTNode
	) => {
		const { line, column: indent } = getPosition(offset, this.lines)

		const children = node.getChildNodes()

		// tslint:disable-next-line: no-shadowed-variable
		function findNode(children) {
			for (let idx = 0; idx < children.length; idx++) {
				const child = children[idx]

				const { line: childLine, column: childCol } = getPosition(
					child.start,
					lines
				)

				if (childCol > indent) {
					return null
				}

				const newChildren = child.getChildNodes()
				const foundNode = findNode(newChildren)

				if (foundNode) {
					return foundNode
				}

				// We have the right indentation, need to return based on line
				if (childLine === line) {
					return child
				}
				if (childLine > line) {
					// Get previous
					return idx - 1 >= 0 ? children[idx - 1] : child
				}
				// Else continue loop to try next element
			}

			// Special case, we found the correct
			return children[children.length - 1]
		}

		return findNode(children) || node
	}
}

function recursivelyBuildAst(parent: ASTNode, node: Yaml.YAMLNode): ASTNode {
	if (!node) {
		return
	}

	switch (node.kind) {
		case Yaml.Kind.MAP: {
			const instance = node as Yaml.YamlMap

			const result = new ObjectASTNode(
				parent,
				null,
				node.startPosition,
				node.endPosition
			)

			for (const mapping of instance.mappings) {
				result.addProperty(recursivelyBuildAst(
					result,
					mapping
				) as PropertyASTNode)
			}

			return result
		}
		case Yaml.Kind.MAPPING: {
			const instance = node as Yaml.YAMLMapping
			const key = instance.key

			// Technically, this is an arbitrary node in YAML
			// I doubt we would get a better string representation by parsing it
			const keyNode = new StringASTNode(
				null,
				null,
				true,
				key.startPosition,
				key.endPosition
			)
			keyNode.value = key.value

			const result = new PropertyASTNode(parent, keyNode)
			result.end = instance.endPosition

			const valueNode = instance.value
				? recursivelyBuildAst(result, instance.value)
				: new NullASTNode(
						parent,
						key.value,
						instance.endPosition,
						instance.endPosition
				  )
			valueNode.location = key.value

			result.setValue(valueNode)

			return result
		}
		case Yaml.Kind.SEQ: {
			const instance = node as Yaml.YAMLSequence

			const result = new ArrayASTNode(
				parent,
				null,
				instance.startPosition,
				instance.endPosition
			)

			let count = 0
			for (const item of instance.items) {
				if (item === null && count === instance.items.length - 1) {
					break
				}

				// Be aware of https://github.com/nodeca/js-yaml/issues/321
				// Cannot simply work around it here because we need to know if we are in Flow or Block
				const itemNode =
					item === null
						? new NullASTNode(
								parent,
								null,
								instance.endPosition,
								instance.endPosition
						  )
						: recursivelyBuildAst(result, item)

				itemNode.location = count++
				result.addItem(itemNode)
			}

			return result
		}
		case Yaml.Kind.SCALAR: {
			const instance = node as Yaml.YAMLScalar
			const type = Yaml.determineScalarType(instance)

			// The name is set either by the sequence or the mapping case.
			const name = null
			const value = instance.value

			// This is a patch for redirecting values with these strings to be boolean nodes because its not supported in the parser.
			const possibleBooleanValues = [
				"y",
				"Y",
				"yes",
				"Yes",
				"YES",
				"n",
				"N",
				"no",
				"No",
				"NO",
				"on",
				"On",
				"ON",
				"off",
				"Off",
				"OFF"
			]
			if (
				instance.plainScalar &&
				possibleBooleanValues.indexOf(value.toString()) !== -1
			) {
				return new BooleanASTNode(
					parent,
					name,
					parseYamlBoolean(value),
					node.startPosition,
					node.endPosition
				)
			}

			switch (type) {
				case Yaml.ScalarType.null: {
					return new StringASTNode(
						parent,
						name,
						false,
						instance.startPosition,
						instance.endPosition
					)
				}
				case Yaml.ScalarType.bool: {
					return new BooleanASTNode(
						parent,
						name,
						Yaml.parseYamlBoolean(value),
						node.startPosition,
						node.endPosition
					)
				}
				case Yaml.ScalarType.int: {
					const result = new NumberASTNode(
						parent,
						name,
						node.startPosition,
						node.endPosition
					)
					result.value = Yaml.parseYamlInteger(value)
					result.isInteger = true
					return result
				}
				case Yaml.ScalarType.float: {
					const result = new NumberASTNode(
						parent,
						name,
						node.startPosition,
						node.endPosition
					)
					result.value = Yaml.parseYamlFloat(value)
					result.isInteger = false
					return result
				}
				case Yaml.ScalarType.string: {
					const result = new StringASTNode(
						parent,
						name,
						false,
						node.startPosition,
						node.endPosition
					)
					result.value = node.value
					return result
				}
			}

			break
		}
		case Yaml.Kind.ANCHOR_REF: {
			const instance = (node as Yaml.YAMLAnchorReference).value

			return (
				recursivelyBuildAst(parent, instance) ||
				new NullASTNode(
					parent,
					null,
					node.startPosition,
					node.endPosition
				)
			)
		}
		case Yaml.Kind.INCLUDE_REF: {
			const result = new StringASTNode(
				parent,
				null,
				false,
				node.startPosition,
				node.endPosition
			)
			result.value = node.value
			return result
		}
	}
}

function convertError(e: Yaml.YAMLException): Problem {
	return {
		message: `${e.reason}`,
		location: {
			start: e.mark.position,
			end: e.mark.position + e.mark.column
		},
		code: ErrorCode.Undefined
	}
}

function createJSONDocument(
	yamlDoc: Yaml.YAMLNode,
	startPositions: number[],
	text: string
) {
	const doc = new SingleYAMLDocument(startPositions)
	doc.root = recursivelyBuildAst(null, yamlDoc)
	doc.globalsConfig = collectGlobals(doc.root)

	if (!doc.root) {
		// TODO: When this is true, consider not pushing the other errors.
		doc.errors.push({
			message: localize(
				"Invalid symbol",
				"Expected a YAML object, array or literal"
			),
			code: ErrorCode.Undefined,
			location: { start: yamlDoc.startPosition, end: yamlDoc.endPosition }
		})

		return doc
	}

	const duplicateKeyReason = "duplicate key"

	// Patch ontop of yaml-ast-parser to disable duplicate key message on merge key
	const isDuplicateAndNotMergeKey = (
		error: Yaml.YAMLException,
		yamlText: string
	) => {
		const errorConverted = convertError(error)
		const errorStart = errorConverted.location.start
		const errorEnd = errorConverted.location.end
		if (
			error.reason === duplicateKeyReason &&
			yamlText.substring(errorStart, errorEnd).startsWith("<<")
		) {
			return false
		}
		return true
	}
	doc.errors = yamlDoc.errors
		.filter(e => e.reason !== duplicateKeyReason && !e.isWarning)
		.map(e => convertError(e))
	doc.warnings = yamlDoc.errors
		.filter(
			e =>
				(e.reason === duplicateKeyReason &&
					isDuplicateAndNotMergeKey(e, text)) ||
				e.isWarning
		)
		.map(e => convertError(e))

	return doc
}

// tslint:disable-next-line: max-classes-per-file
export class YAMLDocument {
	documents: SingleYAMLDocument[]

	constructor(documents: SingleYAMLDocument[]) {
		this.documents = documents
	}
}

export const parse = (text: string, customTags = []): YAMLDocument => {
	const startPositions = getLineStartPositions(text)
	// This is documented to return a YAMLNode even though the
	// typing only returns a YAMLDocument
	const yamlDocs = []

	const schemaWithAdditionalTags = Schema.create(
		customTags.map(tag => {
			const typeInfo = tag.split(" ")
			return new Type(typeInfo[0], { kind: typeInfo[1] || "scalar" })
		})
	)

	// We need compiledTypeMap to be available from schemaWithAdditionalTags before we add the new custom properties
	customTags.map(tag => {
		const typeInfo = tag.split(" ")
		schemaWithAdditionalTags.compiledTypeMap[typeInfo[0]] = new Type(
			typeInfo[0],
			{ kind: typeInfo[1] || "scalar" }
		)
	})

	const additionalOptions: Yaml.LoadOptions = {
		schema: schemaWithAdditionalTags
	}

	Yaml.loadAll(text, docItem => yamlDocs.push(docItem), additionalOptions)

	const doc = new YAMLDocument(
		yamlDocs.map(docItem => {
			const jsonDocument = createJSONDocument(
				docItem,
				startPositions,
				text
			)

			return jsonDocument
		})
	)

	return doc
}
