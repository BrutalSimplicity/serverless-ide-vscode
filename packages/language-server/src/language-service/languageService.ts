import { IConnection } from "vscode-languageserver"
import {
	CompletionList,
	Diagnostic,
	Position,
	TextDocument
} from "vscode-languageserver-types"
import { LanguageSettings } from "./model/settings"
import { YAMLDocument } from "./parser"
import { YAMLCompletion } from "./services/completion"
import { findDocumentSymbols } from "./services/documentSymbols"
import { YAMLHover } from "./services/hover"
import { JSONSchemaService } from "./services/jsonSchema"
import { YAMLValidation } from "./services/validation"

export interface WorkspaceContextService {
	resolveRelativePath(relativePath: string, resource: string): string
}
/**
 * The schema request service is used to fetch schemas. The result should the schema file comment, or,
 * in case of an error, a displayable error string
 */
export type SchemaRequestService = (uri: string) => Promise<string>

export interface CustomFormatterOptions {
	singleQuote?: boolean
	bracketSpacing?: boolean
	proseWrap?: string
}

export interface LanguageService {
	configure(settings: LanguageSettings): void
	doComplete(
		document: TextDocument,
		position: Position,
		doc: YAMLDocument
	): Promise<CompletionList>
	doValidation(document: TextDocument, yamlDocument): Promise<Diagnostic[]>
	doHover(document: TextDocument, position: Position, doc: YAMLDocument)
	findDocumentSymbols(document: TextDocument, doc: YAMLDocument)
	doResolve(completionItem)
}

export function getLanguageService(
	settings: LanguageSettings,
	connection: IConnection
): LanguageService {
	const schemaService = new JSONSchemaService()

	const completer = new YAMLCompletion(schemaService)
	const hover = new YAMLHover(schemaService)
	const yamlValidation = new YAMLValidation(
		schemaService,
		settings.workspaceRoot,
		connection
	)

	const configure = (newSettings: LanguageSettings) => {
		yamlValidation.configure(newSettings)
		hover.configure(newSettings)
		completer.configure(newSettings)
	}

	return {
		configure,
		doComplete: completer.doComplete.bind(completer),
		doResolve: completer.doResolve.bind(completer),
		doValidation: yamlValidation.doValidation.bind(yamlValidation),
		doHover: hover.doHover.bind(hover),
		findDocumentSymbols
	}
}
