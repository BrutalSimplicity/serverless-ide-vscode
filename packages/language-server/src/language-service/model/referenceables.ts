import { PropertyASTNode } from "../parser/json"
import { ReferenceEntityType } from "./references"

export interface Referenceable {
	id: string
	node: PropertyASTNode
	entityType: ReferenceEntityType
}

export interface ReferenceablesHash {
	[key: string]: Referenceable
}

export type ReferenceableLookup = WeakMap<PropertyASTNode, Referenceable>

export interface Referenceables {
	hash: {
		[key in ReferenceEntityType]: ReferenceablesHash
	}
	lookup: ReferenceableLookup
}
