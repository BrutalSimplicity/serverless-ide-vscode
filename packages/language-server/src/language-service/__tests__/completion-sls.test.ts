import { TextDocument } from "vscode-languageserver"
import { getLanguageService } from "../languageService"
import { getDefaultLanguageSettings } from "../model/settings"
import { parse as parseYAML } from "../parser"

const languageSettings = getDefaultLanguageSettings()
const languageService = getLanguageService(languageSettings)

const RESOURCES_TEMPLATE = `
service:
  name: myService
provider:
  name: aws
resources:
  Resources:
    
`

describe("Serverless Framework autocompletion", () => {
	describe("doComplete", () => {
		const setup = (content: string) => {
			return TextDocument.create(
				"file://~/Desktop/serverless.yml",
				"yaml",
				0,
				content
			)
		}

		const parseSetup = async (content: string, offset: number) => {
			const testTextDocument = setup(content)
			const position = testTextDocument.positionAt(offset)

			return await languageService.doComplete(
				testTextDocument,
				position,
				parseYAML(testTextDocument)
			)
		}

		test("should autocomplete resources", async () => {
			const content = RESOURCES_TEMPLATE
			const result = await parseSetup(content, content.length - 1)

			expect(result.items).not.toHaveLength(0)
			expect(result).toMatchSnapshot()
		})
	})
})
