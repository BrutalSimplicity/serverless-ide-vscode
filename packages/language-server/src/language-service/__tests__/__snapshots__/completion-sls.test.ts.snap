// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serverless Framework autocompletion doComplete should autocomplete resources 1`] = `
Object {
  "isIncomplete": false,
  "items": Array [
    Object {
      "documentation": "",
      "insertText": "frameworkVersion: $1",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "frameworkVersion",
      "textEdit": Object {
        "newText": "frameworkVersion: $1",
        "range": Object {
          "end": Object {
            "character": 4,
            "line": 7,
          },
          "start": Object {
            "character": 4,
            "line": 7,
          },
        },
      },
    },
    Object {
      "documentation": "Optional deployment packaging configuration",
      "insertText": "package:
	excludeDevDependencies: \${1:false}",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "package",
      "textEdit": Object {
        "newText": "package:
	excludeDevDependencies: \${1:false}",
        "range": Object {
          "end": Object {
            "character": 4,
            "line": 7,
          },
          "start": Object {
            "character": 4,
            "line": 7,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "functions:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "functions",
      "textEdit": Object {
        "newText": "functions:
	",
        "range": Object {
          "end": Object {
            "character": 4,
            "line": 7,
          },
          "start": Object {
            "character": 4,
            "line": 7,
          },
        },
      },
    },
    Object {
      "documentation": "",
      "insertText": "layers:
	",
      "insertTextFormat": 2,
      "kind": 10,
      "label": "layers",
      "textEdit": Object {
        "newText": "layers:
	",
        "range": Object {
          "end": Object {
            "character": 4,
            "line": 7,
          },
          "start": Object {
            "character": 4,
            "line": 7,
          },
        },
      },
    },
  ],
}
`;
